slicerator.py

```diff
@@ -48,24 +98,101 @@ def __init__(self, ancestor, indices=None, length=None):
         >>> type(v3)
         generator
         """
-        if indices is None:
+        if indices is None and length is None:
             try:
-                indices = range(len(ancestor))
+                length = len(ancestor)
+                indices = range(length)
             except TypeError:
-                raise ValueError("The indices parameter is required in this "
+                raise ValueError("The length parameter is required in this "
                                  "case because len(ancestor) is not valid.")
-        if length is None:
+        elif indices is None:
+            indices = range(length)
+        elif length is None:
             try:
                 length = len(indices)
             except TypeError:
                 raise ValueError("The length parameter is required in this "
                                  "case because len(indices) is not valid.")
+
+        # when list of propagated attributes are given explicitly,
+        # take this list and ignore the class definition
+        if propagated_attrs is not None:
+            self.propagated_attrs = propagated_attrs
+        else:
+            # check propagated_attrs field from the ancestor definition
+            if hasattr(ancestor, 'propagated_attrs'):
+                self.propagated_attrs = ancestor.propagated_attrs
+            else:
+                self.propagated_attrs = []
+
+            # add methods having the _propagate hook
+            for name in dir(ancestor):
+                if hasattr(getattr(ancestor, name), '_propagate'):
+                    self.propagated_attrs.append(name)
+
         self._len = length
         self._ancestor = ancestor
         self._indices = indices
-        self._counter = 0
         self._proc_func = lambda image: image
 
+    @classmethod
+    def from_func(cls, func, length, propagated_attrs=None):
+        """
+        Make a Slicerator from a function that accepts an integer index
+
+        Parameters
+        ----------
+        func : callback
+            callable that accepts an integer as its argument
+        length : int
+            number of elements; used to supposed revserse slicing like [-1]
+        propagated_attrs : list, optional
+            list of attributes to be propagated into Slicerator
+        """
+        class Dummy:
+
+            def __getitem__(self, i):
+                return func(i)
+
+            def __len__(self):
+                return length
+
+        return cls(Dummy(), propagated_attrs=propagated_attrs)
+
+    @classmethod
+    def from_class(cls, other_class, propagated_attrs=None):
```

Comment:
Agreed, and I can't think of an advantage. `isinstance` will still work, which is what matters.